File: javascript/easy/binaryTreePaths.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
var binaryTreePaths = function(root) {
    let results = [], arr = [];
    dfs(root, arr, results);
    return results;
};

let dfs = function(root, arr, results) {
    if(root == null) {
        return;
    } else if(root.left === null && root.right === null) {
        str = "";
        for(let i = 0; i < arr.length; i++) {
            str += arr[i] + "->";
        }
        results.push(str + root.val);
    } else {
        arr.push(root.val);
        dfs(root.left, arr, results);
        dfs(root.right, arr, results);
        arr.pop();
    }
}

File: ruby/easy/binary_tree_paths.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {String[]}
def binary_tree_paths(root)
    results = Array.new
    dfs(root, Array.new, results)
    return results
end

def dfs(root, temp, results)
    return unless root
    if root.left.nil? && root.right.nil?
        str = ""
        temp.each do |i|
            str += i.to_s + "->"
        end
        results.push(str + root.val.to_s)
    else
        temp.push(root.val)
        dfs(root.left, temp, results)
        dfs(root.right, temp, results)
        temp.pop
    end
end
